src/qiskit_qasm3_import/data.py

```diff
@@ -21,6 +21,35 @@ class Scope(enum.Enum):
     NONE = enum.auto()
 
 
+class AddressingMode:
+    """Addressing mode for qubits in OpenQASM 3 programs.
+
+    This class is useful as long as we allow only physical or virtual addressing modes, but
+    not mixed. If the latter is supported in the future, this class will be modified or removed.
+    """
+
+    # 0 == UNKNOWN
+    # 1 == PHYSICAL
+    # 2 == VIRTUAL
+
+    def __init__(self):
+        self._state = 0
+
+    def set_physical_mode(self):
+        """Set the addressing mode to physical. On success return `True`, otherwise `False`."""
+        if self._state != 2:
+            self._state = 1
+            return True
+        return False
+
+    def set_virtual_mode(self):
+        """Set the addressing mode to virtual. On success return `True`, otherwise `False`."""
+        if self._state != 1:
+            self._state = 2
+            return True
+        return False
```

Comment:
The commits are squashed in this PR. But I did use `Enum` earlier: https://github.com/jlapeyre/qiskit-qasm3-import/commit/a4491f4bce9425697f4f4f05d6f499783a0e7ea6 . I didn't like that the details of how we switch modes and what is allowed, and how this is implemented, are spread across a few files and sections of code. I wanted it in one place and behind an abstraction. I noted in the commit message for the current code, https://github.com/jlapeyre/qiskit-qasm3-import/commit/0834755ea5419b0539ceca24349430c14132886c, that the abstraction leaks. But if you compare the code you may agree that it is better this way than with `Enum`. I considered raising the error in `AddressingMode`, which would hide the abstraction at the cost of another call in the call stack. I think that would be a good trade.

I agree with your argument that this does not belong in `data.py`.

Reply:
Personally I think that `Enum` code looks much cleaner, and because it doesn't try to abstract very simple mutation away, it doesn't leak either. If you really did encapsulate the setter and erroring out, I'd use an enum and put the setter logic in `State`, but really I'd prefer to just revert exactly to that commit you had.

Reply:
Let me take another try at a non-leaky abstraction. If it doesn't look good we can revert to Enum. My thought is that with a good abstraction  you have one instantiation of an `AddressingMode` (with no explicit initial data passed). And two calls in two files, one to set physical addressing, the other to set virtual. If there is a problem you get an error.

Compare to `Enum` here: https://github.com/jlapeyre/qiskit-qasm3-import/commit/a4491f4bce9425697f4f4f05d6f499783a0e7ea6 . First, `AddressingMode` should be, as you say, moved, to `converter.py`. But it must be initialized with a value `UNKNOWN`. Then in two different files there are two reads of the data and two mutations. If I have a good abstraction, then I change the implementation, or with luck even the semantics, without changing code at the call sites. But with `Enum` I have to find and understand all five uses of the data and the logic surrounding them.